### ISSUE TITLE
Add Environment.CurrentCoreId API

### ISSUE THREAD

**Issue by benaadams** [dotnet/corefx#696]
## Background

Currently you can have a `ThreadLocal<T>` field or mark something with a `[ThreadStatic]` attribute which can help with concurrency, as unless references are passed around generally only the currently executing thread accessing it.

However, there are high performance, highly parallel and task based situations where what would be more ideal is something focused around cores rather than threads; e.g. a CoreLocal/CoreStatic type.

## Rationale

A common use case for `[ThreadStatic]` is as a caching mechanism as demonstrated in [StringBuilderCache.cs](https://github.com/dotnet/corefx/blob/48d55eece262c82e8be5f8ba37c0ffa3245ae1f2/src/Common/src/System/IO/StringBuilderCache.cs#L47) however in high thread count or task based scenarios this causes undesirable side effects. Core local also helps in high performance situations by making use of the processor cache and in locking & lock-free situations by reducing contention.

### Await fast path

As highlighted by @stephentoub in [ExecutionContext vs SynchronizationContext](http://blogs.msdn.com/b/pfxteam/archive/2012/06/15/executioncontext-vs-synchronizationcontext.aspx) `ThreadLocal` and `ThreadStatic` alter the ambient thread state, which as Tasks can move between threads, the ExecutionContext must be captured and flowed across await points meaning the default ExecutionContext restore fast path is no longer available.

A core focused type would not require the ambient state to be flowed across threads like a thread based types do.

### Memory use

With high thread counts the memory use of  `ThreadLocal` and `ThreadStatic` can become quite high as its a multiple of the thread count. For example using [ConcurrentBag.cs](https://github.com/dotnet/corefx/blob/3f0f482432541f4f2ea7e3f27055cfed1fe071a2/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs#L44) in a high thread situation causes it to become quite a high memory use data structure. While ConcurrentBag is an excellent data structure for a thread focused application; it becomes quite poor in a task focused application.

While thread count is, within reason, unbounded; core count is generally fixed at a far lower level and this would decrease memory use.

### Contention

 `ThreadLocal` and `ThreadStatic` are often used to create variables that can be used without locks/CAS as only one thread can be operating with its thread local variables at a time. A core local variable would have similar effects; though not exactly the same as threads can migrate cores. However, while either locks or CAS semantics would need to be used; core local types would follow a largely uncontended execution path. For example a lock-free algorithm would, more often than not, pass its CAS loop in the first pass even under high load as its based on the core of execution.

### Processor Cache

An advantage that core local would bring over other mechanisms; which is not exposed in the framework is enabling cross Task cache-oblivious algorithms (like `ConcurrentBag` is for threads). Thread core migration aside, a frequently used variable would already be available in the processor cache and it would reduce inter-processor cache synchronisation.

## Framework Use

`ThreadLocal` and `ThreadStatic`  are currently used in the Framework in the current locations: [ConcurrentBag.cs](https://github.com/dotnet/corefx/blob/3f0f482432541f4f2ea7e3f27055cfed1fe071a2/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs#L44), [AllocFreeConcurrentStack.cs](https://github.com/dotnet/corefx/blob/ba8180fdcd863cb6accfb01a98891efe6d989fcd/src/System.Collections.Immutable/src/System/Collections/Immutable/AllocFreeConcurrentStack.cs#L15) and [StringBuilderCache.cs](https://github.com/dotnet/corefx/blob/48d55eece262c82e8be5f8ba37c0ffa3245ae1f2/src/Common/src/System/IO/StringBuilderCache.cs#L47)

### Example Core Local Caching

The following is an example of how a core local approach is possible rather than a proposed api, and shows a change to StringBuilderCache.cs to remove `ThreadStatic` and still achieve the same effect. It's possibly non-ideal as it has both the cost of making a call to kernel32 to determine current processor and a question of portability because of that:

```csharp
using System;
using System.Runtime.InteropServices;
using System.Security;
using System.Text;
using System.Threading;


namespace System.IO
{

    internal static class StringBuilderCache
    {
        private const int MAX_BUILDER_SIZE = 260;
        private const int DEFAULT_CAPACITY = 16;


        [StructLayout(LayoutKind.Explicit, Pack = 1, Size = 64)]
        private struct CacheSpaced<T> where T : class
        {
            // Spaced references to avoid processor cache line sharing
            [FieldOffset(32)]
            public T Value;

        }

        private static CacheSpaced<StringBuilder>[] t_cachedInstances;

        [DllImport("kernel32.dll"), SuppressUnmanagedCodeSecurity]
        private static extern int GetCurrentProcessorNumber();

        public static StringBuilderCache()
        {
            t_cachedInstances = new CacheSpaced<StringBuilder>[Environment.ProcessorCount];
        }

        public static StringBuilder Acquire(int capacity = DEFAULT_CAPACITY)
        {
            if (capacity <= MAX_BUILDER_SIZE)
            {
                StringBuilder sb = Interlocked.Exchange(ref StringBuilderCache.t_cachedInstances[GetCurrentProcessorNumber()].Value, null);
                if (sb != null)
                {
                    // Avoid stringbuilder block fragmentation by getting a new StringBuilder
                    // when the requested size is larger than the current capacity
                    if (capacity <= sb.Capacity)
                    {
                        sb.Clear();
                        return sb;
                    }
                    else
                    {
                        // Capacity not large enough, put back to cache
                        Interlocked.Exchange(ref StringBuilderCache.t_cachedInstances[GetCurrentProcessorNumber()].Value, sb);
                    }
                }
            }
            return new StringBuilder(capacity);
        }

        public static StringBuilder Acquire(string value)
        {
            StringBuilder sb = Acquire(Math.Max(value.Length, DEFAULT_CAPACITY));
            sb.Append(value);
            return sb;
        }

        public static void Release(StringBuilder sb)
        {
            if (sb.Capacity <= MAX_BUILDER_SIZE)
            {
                Interlocked.Exchange(ref StringBuilderCache.t_cachedInstances[GetCurrentProcessorNumber()].Value, sb);
            }
        }

        public static string GetStringAndRelease(StringBuilder sb)
        {
            string result = sb.ToString();
            Release(sb);
            return result;
        }
    }
}
```

## Open Questions

Would be purely corefx or require coreclr backing? As it either requires the asm execution  of RDTSCP or kernel32.dll's GetCurrentProcessorNumber.

Cross platform would need different internal semantics as GetCurrentProcessorNumber is Windows and RDTSCP is x86

How to communicate usage to be cautious that executing threads can migrate cores?

**Comment by svick** [comment]
> in high thread count or task based scenarios this causes undesirable side effects

Are such scenarios actually common? Using `Task`s and `async`-`await` should mean that the thread count is not high, so core-local data might not help that much.

Do you have measurements of realistic situations that indicate this would help enough to be worth it?

> A core focused type would not require the ambient state to be flowed across threads like a thread based types do.

How come? The state must be flowed currently, because an `await`-based `Task` can move between threads. But all `Task`s can move between cores, so `Task`-local data can't be kept in core-local storage either.

>  cache-oblivious algorithms

This term has [a very specific meaning](https://en.wikipedia.org/wiki/Cache-oblivious_algorithm), I don't think that's what you meant here.

**Comment by stephentoub** [comment]
Thanks for the suggestion.

> A core focused type would not require the ambient state to be flowed across threads like a thread based types do.

Why?  The reason this state needs to be flowed is that the async operation may not continue on the same thread on which it started, and that in the meantime state on that thread could have been changed.  What would guarantee that the async operation would continue on the same core on which it started, or that even if it did the state wouldn't have been changed in the interim?

> Thread core migration aside

That's a big "aside" ;)

> there are high performance, highly parallel and task based situations where what would be more ideal is something focused around cores rather than threads;

There are, but they'd really require more than such data types.  They'd also benefit from thread affinitization, especially if they want to make any guarantees on state being in their "core local" or "core static" variable from one moment to the next, and without such guarantees, such types could end up being really dangerous.  Thread affinitization is also something that general-purpose libraries really shouldn't be doing and enforcing on their consuming apps (or other libraries), which means such types would likely be limited to consumption at the app-level where control over thread affinity is more reasonable.

I agree that there are some version niche cases where such support would be useful.  But I don't think such types should be provided in a general-purpose library.

**Comment by benaadams** [comment]
> How come? The state must be flowed currently, because...

and

> The reason this state needs to be flowed is that the async operation may not continue on the same thread on which it started, and that in the meantime state on that thread could have been changed. What would guarantee that the async operation would continue on the same core...

In a lot of the situations you don't really care, which why I included the change to StringBuilderCache to demonstrate this. A use of `ThreadStatic` for caching doesn't particularly care about the thread state an would use a flow similar to:
```csharp
// get from cache and null cache
T local = Interlocked.Exchange(ref cached, null);
if (local == null)  {
    local = new T();
}
// do work ...
Interlocked.Exchange(ref cached, local); // put back in cache
```
so only partial semantics of `ThreadStatic` are needed; however because `ThreadStatic` has other guarantees using it must invoke full ambient state flow. In the current situation if my code uses StringBuilder somewhere and I use tasks elsewhere; say in a server app, I've probably killed all the fast path ExecutionContext restores because in a short amount of time all my threads now have a `[ThreadStatic] StringBuilder` attached to them which must be flowed. (Might be wrong about this?) I'm not saying the flowing is incorrect, it is very correct; however in this situation it is an unnecessary side effect; which not using `ThreadStatic` would solve.

> They'd also benefit from thread affinitization...

if you want to open that box ;-) However, that's a different area that requires fixed threads of execution, when class variables would cover this case; and you probably won't want to use any tasks.

**Comment by stephentoub** [comment]
>  because ThreadStatic has other guarantees using it must invoke full ambient state flow. In the current situation if my code uses StringBuilder somewhere and I use tasks elsewhere; say in a server app, I've probably killed all the fast path ExecutionContext restores because in a short amount of time all my threads now have a [ThreadStatic] StringBuilder attached to them which must be flowed. (Might be wrong about this?

TLS (whether [ThreadStatic] or ThreadLocal) is not actually part of ExecutionContext and isn't flowed with it.  My "this state" was referring to the ExecutionContext.  I should have been more clear in my initial response.  I was just trying to point out that your argument around a core-local state being better for async methods than thread-local state doesn't really pan out; if the state is not flowed, then it doesn't matter in either case, and if the state does need to be flowed, then it would have to be for core-local as well.

> In a lot of the situations you don't really care, which why I included the change to StringBuilderCache to demonstrate this.

Yes, but you have to be incredibly careful when you use such state.  You were careful in your order of operations and with CAS instructions to make sure that no one else could have used the state you grabbed, even if you migrated cores and even if you grabbed the state associated with the wrong core.  If that logic were not built into such a type, then we'd be giving people an easily-misused construct, and if it were built-in, it'd further limit the applicability of what I'd argue is already a very niche type.

Also, in your example with StringBuilder, how much cache efficiency are you really getting?  You're potentially migrating cores while using it, you're potentially putting it back on a different core than it came from, etc.  I'm curious... have you done measurements with your example to show that the wins exist and are measurable?

**Comment by stephentoub** [comment]
Also, I didn't understand the ConcurrentBag example.  ConcurrentBag is designed such that each thread has its own thread-local list, such that each thread can make assumptions about the synchronization required to access that list.  How do you see that playing out if it were to use your core-local example, if thread affinity is not used?  Or was that not what you were suggesting?

**Comment by benaadams** [comment]
> But all Tasks can move between cores, so Task-local data can't be kept in core-local storage either.

I'm not suggesting `Task`-local data; that essentially already there in the regular execution flow or local variables hoisted to the state machine. Its `Thread`-global variables from `ThreadStatic` and `ThreadLocal` that cause the problem.

> Are such scenarios actually common? 

They are for me, though I can't speak for the wider community. Our use case is in high throughput async websockets as used by https://msdn.microsoft.com/en-us/age-of-ascent-msdn 

> This term has a very specific meaning, I don't think that's what you meant here.

`ConcurrentBag` for example has some really nice algorithms that are very similar so what I am suggesting; operating on local lists, then stealing from other lists when the local lists are done. However its thread focused and doesn't work as well in a high thoughput task focused situation when `ConcurrentList` actually works better; however `ConcurrentBag` has some really smart algorithms but its the relationship between Tasks and Threads that causes this mismatch.

**Comment by benaadams** [comment]
> I'm curious... have you done measurements with your example to show that the wins exist and are measurable

I do have a microbenchmark I'll see if I can upload to github; been a little while since I looked at it. With the call to kernel32.dll it starts off more expensive in low contention/threading, but scales better as contention/threading increases. I don't have an example for ThreadStatic as that would be a bit of weird one to measure but might be able to contrive something :)

> ConcurrentBag is designed such that each thread has its own thread-local list, such that each thread can make assumptions about the synchronization required to access that list. How do you see that playing out if it were to use your core-local example, if thread affinity is not used?

In [ConcurrentBag](https://github.com/dotnet/corefx/blob/3f0f482432541f4f2ea7e3f27055cfed1fe071a2/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs#L276) essentially the same approach but based around something like a `Environment.CurrentCoreId` rather than `Environment.CurrentManagedThreadId`. Essentially `ConcurrentBag` is the exact kind of data structure that would work if you needed to hide the difficulties and pitfalls; and expose a fast `Environment.CurrentCoreId`for the foolhardy ;-)

**Comment by stephentoub** [comment]
> In ConcurrentBag essentially the same approach but based around something like a Environment.CurrentCoreId rather than Environment.CurrentManagedThreadId. Essentially ConcurrentBag is the exact kind of data structure that would work if you needed to hide the difficulties and pitfalls; and expose a fast Environment.CurrentCoreIdfor the foolhardy ;-)

I think this is a good example of where this breaks down, though.

The work-stealing algorithms employed by ConcurrentBag allow a thread to operate on its own thread-local list without locks in some cases, such as in the case of adding to its own list when it has at least [two elements](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs#L128).  If its list were simply changed to be a core-local rather than thread-local list, this becomes a bug, as it's now possible for multiple threads to mutate a list they believe they own and without any concurrency.  If you instead follow the approach of atomically removing the list so that no one else can use it, a) that adds to the synchronization costs, but more importantly b) then what?  How does the list get put back?  If the thread migrates to another core that already has a list, what does it do?  It can't just throw it away nor overwrite the one that's there, as the data is important.  So it would have to somehow get the data from this list added into the existing list, while ensuring that any stealers are appropriately synchronized, and so on.  Or maybe the entire work-stealing algorithm would have to be overhauled.

Alternatively, you could imagine using thread affinitization.  If such affinitization could be guaranteed for all threads using the bag, then a core-local list would work, but then again it'd have the same performance as a thread-local list, since at that point a thread would be tied 1:1 to a particular core.  And if instead the thread affinity were provided by the library, every operation would require checking for the current affinity, setting the new affinity, and resetting back to the original affinity, which brings it own overheads and issues.

Again, I appreciate the suggestion, and I agree there are places where this kind of support could be beneficial for very particular kinds of workloads, but I really worry that exposing such functionality in a general-purpose library will lead to way more problems than it solves.

> I do have a microbenchmark I'll see if I can upload to github

Thanks.

**Comment by benaadams** [comment]
Microbenchmark still runs :) Results as follows:

> 10,000,000 increments per thread via function call
Conc: Speed improvement vs 1 thread.
Scale: Threading improvement vs ideal.

> Unsync'd raw variable
[1 thread ] ave 234.6ms, min 68.0ms, max 863.0ms
[2 threads] ave 277.3ms, min 183.0ms, max 587.0ms, conc 1.7x, scale 85%
[4 threads] ave 599.9ms, min 304.0ms, max 1,264.0ms, conc 1.6x, scale 39%
[8 threads] ave 1,283.9ms, min 843.0ms, max 1,371.0ms, conc 1.5x, scale 18%
Interlocked.Increment on variable
[1 thread ] ave 534.0ms, min 132.0ms, max 2,142.0ms
[2 threads] ave 1,046.0ms, min 785.0ms, max 1,665.0ms, conc 1.0x, scale 51%
[4 threads] ave 1,788.1ms, min 1,470.0ms, max 2,672.0ms, conc 1.2x, scale 30%
[8 threads] ave 2,832.3ms, min 943.0ms, max 3,578.0ms, conc 1.5x, scale 19%
Increment++ inside lock()
[1 thread ] ave 670.4ms, min 348.0ms, max 1,956.0ms
[2 threads] ave 951.6ms, min 576.0ms, max 2,231.0ms, conc 1.4x, scale 70%
[4 threads] ave 1,445.4ms, min 940.0ms, max 2,668.0ms, conc 1.9x, scale 46%
[8 threads] ave 3,401.3ms, min 2,155.0ms, max 3,975.0ms, conc 1.6x, scale 20%
Core-local (Counter)
[1 thread ] ave 397.6ms, min 286.0ms, max 841.0ms
[2 threads] ave 382.6ms, min 287.0ms, max 716.0ms, conc 2.1x, scale 104%
[4 threads] ave 377.4ms, min 299.0ms, max 583.0ms, conc 4.2x, scale 105%
[8 threads] ave 579.7ms, min 472.0ms, max 1,327.0ms, conc 5.5x, scale 69%
Core-local + Interlocked.Increment (Counter Safe)
[1 thread ] ave 516.8ms, min 319.0ms, max 1,088.0ms
[2 threads] ave 636.6ms, min 322.0ms, max 1,414.0ms, conc 1.6x, scale 81%
[4 threads] ave 730.6ms, min 365.0ms, max 2,526.0ms, conc 2.8x, scale 71%
[8 threads] ave 1,067.6ms, min 867.0ms, max 2,368.0ms, conc 3.9x, scale 48%

> Performance order fastest to slowest:
[1 thread ] Unsync'd (x1.0), Counter (x1.7), CounterSafe (x2.2), Interlocked (x2.3), Locked (x2.9)
[2 thread ] Unsync'd (x1.0), Counter (x1.4), CounterSafe (x2.3), Locked (x3.4), Interlocked (x3.8)
[4 thread ] Counter (x1.0), Unsync'd (x1.6), CounterSafe (x1.9), Locked (x3.8), Interlocked (x4.7)
[8 thread ] Counter (x1.0), CounterSafe (x1.8), Unsync'd (x2.2), Interlocked (x4.9), Locked (x5.9)

> Validation [All values should be the same]:
(-72.7% Error) Unsync'd value: 432079822
(0.0% Error) Interlocked value 1580000000
(0.0% Error) Locked value 1580000000
(0.0% Error) Counter value 1579999885
(0.0% Error) Counter Safe value 1580000000

Run 10,000,000 increments per thread + warm up making 1,580,000,000 increments. 

The unsynchronised value performs very badly beyond 1 thread and the result is out by 72.7% (1,147,920,178) due to the race conditions. 

The Counter (core-local) is out by 115 as again there is no synchronization and this will be caused by core swaps. Using Interlocked to do the increment fixes the result; but doesn't suffer as much as just using Interlocked on a single value as the increments are mostly cache uncontended.

Will put some code up...

W

**Comment by benaadams** [comment]
Might be helpful if there were some context!

4 physical core machine, 8 logical (HT) Intel Core i7 - 3632QM @ 2.20Ghz so don't have 8 cores for the 8 threads

**Comment by omariom** [comment]
Quite often when dealing with core local data structures one has to cache align them - another concern for users of such feature to take into account.

Imo, currently CLR provides good enough mechanisms to implement them.  The only question, as @benaadams noted, is portability. If CoreFx could provide us with portable way of getting current core number and as performant as GetCurrentProcessorNumber (though it is much slower in x86 process) that would be enough.

**Comment by benaadams** [comment]
@omariom agreed, would be happy with something like Environment.CurrentCoreId

**Comment by stephentoub** [comment]
@benaadams, @omariom, something like Environment.CurrentCoreId is a reasonable request.  We'll add it to the list of API additions to consider.  Thanks for the discussion.

(@benaadams, I'd still be interested in seeing the code for your microbenchmark... it also looks like it's missing a TLS case for comparison purposes, but maybe I've misunderstood what the various cases are.)

**Comment by benaadams** [comment]
@stephentoub there are a couple things missing for this particular issue, its from a cache effects benchmark I did in 2013 for use in a Task based system; my coding isn't that fast! :) 

TLS would probably do very well for the thread tests, I assume as well as the core-local test, if not better. Is just for a Task approach TLS becomes a bit weird it makes the abstraction become leaky as the threads become relevant. Will add it though and see if I can add some Task based tests.

**Comment by benaadams** [comment]
@stephentoub from the "code never meant to be seen" school of development: https://github.com/benaadams/CoreLocalTest apologies for the quality!

**Comment by benaadams** [comment]
Updated [with TLS, updated sample to x64 rather than x86](https://github.com/benaadams/CoreLocalTest) as there is a speed difference, increased repeats, decreased tight loop size.

> Increment value tests
Conc: Speed improvement vs 1 thread.
Scale: Threading improvement vs ideal.
Benchmarking Unsync'd raw variable
[1 thread ] ave 6.9ms, min 6.0ms, max 70.0ms
[2 threads] ave 22.6ms, min 11.0ms, max 63.0ms, conc 0.6x, scale 30%
[4 threads] ave 42.7ms, min 27.0ms, max 122.0ms, conc 0.6x, scale 16%
[8 threads] ave 129.2ms, min 57.0ms, max 143.0ms, conc 0.4x, scale 5%
Benchmarking Interlocked.Increment on variable
[1 thread ] ave 15.3ms, min 13.0ms, max 229.0ms
[2 threads] ave 69.5ms, min 19.0ms, max 157.0ms, conc 0.4x, scale 22%
[4 threads] ave 140.7ms, min 105.0ms, max 281.0ms, conc 0.4x, scale 11%
[8 threads] ave 303.6ms, min 144.0ms, max 545.0ms, conc 0.4x, scale 5%
Benchmarking Increment++ inside lock()
[1 thread ] ave 37.9ms, min 34.0ms, max 317.0ms
[2 threads] ave 65.8ms, min 35.0ms, max 517.0ms, conc 1.2x, scale 58%
[4 threads] ave 123.5ms, min 39.0ms, max 453.0ms, conc 1.2x, scale 31%
[8 threads] ave 311.2ms, min 138.0ms, max 383.0ms, conc 1.0x, scale 12%
Benchmarking Core-local (Counter)
[1 thread ] ave 28.1ms, min 28.0ms, max 40.0ms
[2 threads] ave 47.2ms, min 28.0ms, max 83.0ms, conc 1.2x, scale 60%
[4 threads] ave 53.5ms, min 28.0ms, max 77.0ms, conc 2.1x, scale 53%
[8 threads] ave 79.9ms, min 77.0ms, max 116.0ms, conc 2.8x, scale 35%
Benchmarking Core-local + Interlocked.Increment (Counter Safe)
[1 thread ] ave 32.1ms, min 31.0ms, max 37.0ms
[2 threads] ave 32.7ms, min 31.0ms, max 52.0ms, conc 2.0x, scale 98%
[4 threads] ave 43.0ms, min 31.0ms, max 55.0ms, conc 3.0x, scale 75%
[8 threads] ave 53.6ms, min 51.0ms, max 57.0ms, conc 4.8x, scale 60%
Benchmarking TLS Increment
[1 thread ] ave 40.1ms, min 38.0ms, max 84.0ms
[2 threads] ave 41.5ms, min 38.0ms, max 140.0ms, conc 1.9x, scale 97%
[4 threads] ave 60.1ms, min 38.0ms, max 95.0ms, conc 2.7x, scale 67%
[8 threads] ave 84.5ms, min 70.0ms, max 126.0ms, conc 3.8x, scale 47%
Benchmarking TLS cache-isolated Increment
[1 thread ] ave 32.3ms, min 27.0ms, max 37.0ms
[2 threads] ave 34.9ms, min 27.0ms, max 59.0ms, conc 1.9x, scale 93%
[4 threads] ave 44.2ms, min 27.0ms, max 60.0ms, conc 2.9x, scale 73%
[8 threads] ave 54.2ms, min 47.0ms, max 74.0ms, conc 4.8x, scale 60%

> Perfromance order fastest to slowest:
**[1 thread ]** Unsync'd raw variable (x1.0), Interlocked.Increment on variable (x2.2), Core-local (Counter) (x4.1), **Core-local + Interlocked.Increment (Counter Safe) (x4.7)**, **TLS cache-isolated Increment (x4.7)**, Increment++ inside lock() (x5.5), TLS Increment (x5.8)

> **[2 thread ]** Unsync'd raw variable (x1.0), **Core-local + Interlocked.Increment (Counter Safe) (x1.5)**, **TLS cache-isolated Increment (x1.5)**, TLS Increment (x1.8), Core-local (Counter) (x2.1), Increment++ inside lock() (x2.9), Interlocked.Increment on variable (x3.1)

> **[4 thread ]** Unsync'd raw variable (x1.0), **Core-local + Interlocked.Increment (Counter Safe) (x1.0)**, **TLS cache-isolated Increment (x1.0)**, Core-local (Counter) (x1.3), TLS Increment (x1.4), Increment++ inside lock() (x2.9), Interlocked.Increment on variable (x3.3)

> **[8 thread ]** **Core-local + Interlocked.Increment (Counter Safe) (x1.0)**, **TLS cache-isolated Increment (x1.0)**, Core-local (Counter) (x1.5), TLS Increment (x1.6), Unsync'd raw variable (x2.4), Interlocked.Increment on variable (x5.7), Increment++ inside lock() (x5.8)

Turns out I have misunderstood something. @stephentoub as you pointed out earlier and as I have only just realised is TLS doesn’t flow across async points. In which case the advantages reduce to slightly reduced memory and some processor cache optimisations for specialist algorithms - at which point Environment.CurrentCoreId and Affinities are the way to go...

**Comment by stephentoub** [comment]
Thanks, @benaadams.  I've reopened the issue and renamed it to help track the API request.

Thanks for sharing your microbenchmark code.  I totally understand the "code never meant to be seen", so I didn't spend a lot of time looking over it, but one thing did pop out at me that could affect the results.  You're using Task.WhenAll(t1, t2, ...) to wait on your tasks, but WhenAll just returns a Task that will complete when all of the consistuent tasks complete, so this isn't actually waiting for anything... you need to either await or Wait on the Task returned from WhenAll, or you need to use Task.WaitAll instead.

Even with that, from the results you posted, there's basically no perf difference between the core-local and thread-local tests.  In support of arguing for adding an Environment.CurrentCoreId-like member, can you share an example that does measurably benefit?  A new API like this has some downsides (beyond the costs inherent to adding any new API), namely the issues we've discussed of a developer really needing to understand that it can change immediately after it's been accessed, so we'd need some really good examples that demonstrate why its value outweighs its risks.

Thanks!

**Comment by benaadams** [comment]
Whoops, fixed that. Largely this request has arisen from a misunderstanding I had about the interaction of TLS and await; specifically in the previously linked blog post:

> **ExecutionContext is all about “ambient” information**, meaning that it stores data relevant to the current environment or “context” in which you’re running.  In many systems, **such ambient information is maintained in thread-local storage (TLS)**, such as in a ThreadStatic field or in a ThreadLocal<T>. 

but I then missed this bit

> because TLS **doesn’t** “flow” across these async points. 

So really TLS fits my needs, which is why I closed the issue. Also contrary to my earlier suggestion I'd suggest it should be in System.Diagnostics rather than Environment. 

Around this area, the thread affinity apis could do with some work. I might not be correct on this but as I understand it, on windows to change the affinity for a managed thread first you want to bind it: [Thread.BeginThreadAffinity](https://msdn.microsoft.com/en-us/library/system.threading.thread.beginthreadaffinity(v=vs.110).aspx) then dive through kernel32's GetCurrentThread and SetThreadAffinityMask. However, you can change a different processes thread's affinities via [ProcessThread.IdealProcessor](https://msdn.microsoft.com/en-us/library/system.diagnostics.processthread.idealprocessor(v=vs.110).aspx) and [ProcessThread.ProcessorAffinity](https://msdn.microsoft.com/en-us/library/system.diagnostics.processthread.processoraffinity(v=vs.110).aspx). On Xbox360 there is [Thread.SetProcessorAffinity](https://msdn.microsoft.com/en-us/library/system.threading.thread.setprocessoraffinity(v=vs.110).aspx). On linux would it be forked processes etc?

If you set thread affinities, then you should already know the processor, so CurrentCoreId might not be needed. Also good processor cache aware algos are generally in tight loops rather than over time; and with thread affinities, its also very easy to do more harm than good... though I could see uses in single purpose devices/appliances/IoT or dealing with NUMA nodes etc... 

Long story short, I cannot provide a good justification within my field.

**Comment by stephentoub** [comment]
Thanks for the aditional follow-up, @benaadams.  In that case, I agree, we should close this out.  Thanks for the discussion!
